smarts/core/waymo_map.py

```diff
@@ -234,6 +245,98 @@ def _raycast_boundaries(
 
         return left_widths, right_widths
 
+    def _compute_lane_intersections(self) -> Dict[int, Set[int]]:
+        all_lane_ids = [feat_id for feat_id in self._feat_dicts.keys()]
+        intersections: Dict[int, Set[int]] = dict()
+
+        # Set up the intersections dict
+        for feat_id in all_lane_ids:
+            intersections[feat_id] = set()
+
+        # Build rtree
+        lane_rtree = rtree.index.Index()
+        lane_rtree.interleaved = True
+        bboxes = dict()
+        for idx, feat_id in enumerate(all_lane_ids):
+            lane_pts = np.array(self._polyline_cache[feat_id][0])
+            bbox = BoundingBox(
+                min_pt=Point(x=np.amin(lane_pts[:, 0]), y=np.amin(lane_pts[:, 1])),
+                max_pt=Point(x=np.amax(lane_pts[:, 0]), y=np.amax(lane_pts[:, 1])),
+            )
+            bboxes[feat_id] = bbox
+
+            coords = (
+                bbox.min_pt.x,
+                bbox.min_pt.y,
+                bbox.max_pt.x,
+                bbox.max_pt.y,
+            )
+            lane_rtree.add(idx, coords)
+
+        # Loop over every lane in the map
+        for feat_id in all_lane_ids:
+            # Filter out any lanes that don't intersect this lane's bbox
+            bbox = bboxes[feat_id]
+            indicies = lane_rtree.intersection(
+                (
+                    bbox.min_pt.x,
+                    bbox.min_pt.y,
+                    bbox.max_pt.x,
+                    bbox.max_pt.y,
+                )
+            )
+
+            # Filter out any other lanes we don't want to check against
+            lanes_to_test = []
+            for idx in indicies:
+                cand_id = all_lane_ids[idx]
+                features = self._feat_dicts[feat_id]
+
+                # Don't check intersection with incoming/outgoing lanes or itself
+                in_ids = [l for l in features["entry_lanes"]]
+                out_ids = [l for l in features["exit_lanes"]]
+                if cand_id in in_ids + out_ids + [feat_id]:
+                    continue
+
+                # Skip intersections we've already computed
+                if cand_id in intersections[feat_id]:
+                    continue
+
+                lanes_to_test.append(cand_id)
+
+            # Main loop -- check each segment of the lane polyline against the
+            # polyline of each candidate lane
+            line1 = np.array(self._polyline_cache[feat_id][0])
+            for cand_id in lanes_to_test:
+                line2 = np.array(self._polyline_cache[cand_id][0])
+                C = np.roll(line2, 0, axis=0)[:-1]
+                D = np.roll(line2, -1, axis=0)[:-1]
+                len_c = len(C)
+                for i in range(len(line1) - 1):
+                    A = np.tile(line1[i], (len_c, 1))
+                    B = np.tile(line1[i + 1], (len_c, 1))
```

Comment:
Re our discussion about how to avoid the repeated allocations from `np.tile()` here, I think this page has answers:  https://numpy.org/doc/stable/user/basics.broadcasting.html

Long story short, I don't think you need the `tile()` calls here.

Reply:
Yes you're right -- good catch! It looks like this does indeed give us an additional 10-20% speedup too.